File: node-bun/src/commands/moderation/case.js
import { resolveMember, basicEmbed, error } from '../../utils/helpers.js';
import { checkMod } from '../../utils/checks.js';
import { getCase, getCasesForUser, getRecentCases } from '../../database/db.js';
import { formatDate } from '../../utils/time.js';

function formatCaseExtra(extra = {}) {
  const lines = [];
  if (extra.source) lines.push(`**Source:** ${extra.source}`);
  if (extra.status) lines.push(`**Status:** ${extra.status}`);
  if (extra.warning_id) lines.push(`**Warning ID:** ${extra.warning_id}`);
  if (extra.queue_id) lines.push(`**Queue ID:** ${extra.queue_id}`);
  if (extra.timed_action_id) lines.push(`**Timed action ID:** ${extra.timed_action_id}`);
  if (extra.timed_action) lines.push(`**Timed action:** ${extra.timed_action}`);
  if (extra.ends_at) lines.push(`**Expires:** ${formatDate(extra.ends_at)}`);
  return lines;
}

export default {
  name: 'case',
  description: 'View moderation cases by number or user',
  category: 'Moderation',
  usage: '?case <number> or ?case list [user]',
  async execute(message, args, subcommand, subargs) {
    const denied = await checkMod(message);
    if (denied) return message.reply(denied);

    const action = subcommand?.toLowerCase();
    const rest = subargs || args.trim();

    if (action === 'list') {
      const target = resolveMember(message, rest.trim());
      const cases = target
        ? await getCasesForUser(message.guild.id, target.id)
        : await getRecentCases(message.guild.id);

      if (!cases.length) return message.reply(error('No cases found.'));

      const title = target ? `Cases: ${target.user.tag}` : 'Recent Cases';
      const lines = cases.map((c) =>
        `**#${c.case_number}** ${c.action} — ${c.reason} by <@${c.moderator_id}> (${formatDate(c.created_at)})`
      );
      return message.reply({ embeds: [basicEmbed(title, lines.join('\n'))] });
    }

    const caseNum = parseInt(action && /^\d+$/.test(action) ? action : rest.trim(), 10);
    if (!caseNum) return message.reply(error('Usage: `?case <number>` or `?case list [user]`'));

    const caseData = await getCase(message.guild.id, caseNum);
    if (!caseData) return message.reply(error(`Case #${caseNum} not found.`));

    const embed = basicEmbed(`Case #${caseData.case_number}`, [
      `**Action:** ${caseData.action}`,
      `**User:** <@${caseData.user_id}> (\`${caseData.user_id}\`)`,
      `**Moderator:** <@${caseData.moderator_id}>`,
      `**Reason:** ${caseData.reason || 'None'}`,
      `**Date:** ${formatDate(caseData.created_at)}`,
      ...formatCaseExtra(caseData.extra),
    ].join('\n'));

    return message.reply({ embeds: [embed] });
  },
};

File: python/bot/cogs/moderation/case.py
"""?case — moderation case history."""

from __future__ import annotations

from discord.ext import commands

from bot.checks.decorators import moderator_required
from bot.cogs.deps import CogRepos
from bot.utils.helpers import basic_embed, error
from bot.utils.resolvers import resolve_member
from bot.utils.time import format_iso_date


def _format_case_extra(metadata: dict) -> list[str]:
    lines: list[str] = []
    for key, label in (
        ("source", "Source"),
        ("status", "Status"),
        ("warning_id", "Warning ID"),
        ("queue_id", "Queue ID"),
        ("timed_action_id", "Timed action ID"),
        ("timed_action", "Timed action"),
        ("ends_at", "Expires"),
    ):
        if key in metadata and metadata[key] is not None:
            value = metadata[key]
            if key == "ends_at" and isinstance(value, int):
                lines.append(f"**{label}:** <t:{value // 1000}:f>")
            else:
                lines.append(f"**{label}:** {value}")
    return lines


class CaseCog(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot
        self.repos = CogRepos(bot)

    @commands.group(name="case", invoke_without_command=True)
    @moderator_required()
    async def case(self, ctx: commands.Context[commands.Bot], *, arg: str = "") -> None:
        if ctx.invoked_subcommand is not None:
            return
        trimmed = arg.strip()
        if not trimmed:
            await ctx.reply(error("Usage: `?case <number>` or `?case list [user]`"))
            return
        try:
            case_number = int(trimmed.replace("#", ""), 10)
        except ValueError:
            await ctx.reply(error("Usage: `?case <number>` or `?case list [user]`"))
            return
        await self._show_case(ctx, case_number)

    async def _show_case(self, ctx: commands.Context[commands.Bot], case_number: int) -> None:
        record = await self.repos.cases.get_case(str(ctx.guild.id), case_number)  # type: ignore[union-attr]
        if record is None:
            await ctx.reply(error(f"Case #{case_number} not found."))
            return
        body = "\n".join(
            [
                f"**Action:** {record.action}",
                f"**User:** <@{record.user_id}> (`{record.user_id}`)",
                f"**Moderator:** <@{record.moderator_id}>",
                f"**Reason:** {record.reason or 'None'}",
                f"**Date:** {format_iso_date(record.created_at)}",
                *_format_case_extra(record.metadata),
            ],
        )
        await ctx.reply(embed=basic_embed(f"Case #{record.case_number}", body))

    @case.command(name="list")
    @moderator_required()
    async def case_list(self, ctx: commands.Context[commands.Bot], *, user_arg: str | None = None) -> None:
        guild_id = str(ctx.guild.id)  # type: ignore[union-attr]
        target = resolve_member(ctx.guild, ctx.message, user_arg)  # type: ignore[arg-type]
        cases = (
            await self.repos.cases.list_for_user(guild_id, str(target.id))
            if target
            else await self.repos.cases.list_recent(guild_id)
        )
        if not cases:
            await ctx.reply(error("No cases found."))
            return
        title = f"Cases: {target.display_name}" if target else "Recent Cases"
        lines = [
            f"**#{c.case_number}** {c.action} — {c.reason} by <@{c.moderator_id}> ({format_iso_date(c.created_at)})"
            for c in cases
        ]
        await ctx.reply(embed=basic_embed(title, "\n".join(lines)))


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(CaseCog(bot))
